<a href="https://colab.research.google.com/github/cata603/An-litica-de-negocios/blob/main/Ejercicio_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Importar las librerías necesarias

In [ ]:
import numpy as np  # Contiene funciones numéricas de transformación y análisis de datos
import pandas as pd  # Sirve para conectar el sistema con las bases de datos
from sklearn.naive_bayes import GaussianNB  # Algoritmo Naive Bayes, usado para clasificación

1. Cargar la base de datos de accidentes desde el archivo Excel

In [ ]:
nxl = "/content/1. BD2_CarCrash_Int.xlsx"  # Ruta del archivo
XDB = pd.read_excel(nxl, sheet_name=0)  # Para leer los datos de la primera hoja del archivo Excel

2. Seleccionar las columnas relevantes para el análisis y filtrar

3. Separar las variables de entrada y salida y apicar el metodo Naive bayes:

In [ ]:
X = XDB1[['Weekday', 'ViolCat', 'ClearWeather', 'Month', 'CrashType', 'Daylight']]  # Variables explicativas
y = XDB1['Highway']  # Variable objetivo (si ocurrió en una autopista o no)


mnb=GaussianNB()  #mnb= model naive bayes
mnb.fit(X,y)   # Para entrenar el modelo usando las variables

ct=mnb.class_count_  # Muestra la distribución de las clases en el modelo entrenado
print ("Como estan separados los datos", ct)
print ("El total de los datos es", len(X))
print ("Los datos porcentuales estab div:", ct/len(X))

#Para evaluar un caso puntual
XS=[[5,1,1,3,4,0]]
XS= mnb.predict(XS)
print(XS)


Como estan separados los datos [1580.  501.]
El total de los datos es 2081
Los datos porcentuales estab div: [0.75925036 0.24074964]
[0]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


4. Hallar datos relevantes (media, desviación, limites superiores e inferiores)

In [ ]:
u = mnb.theta_  # Media de cada variable para cada categoría de 'Highway'
print("La media de las variables por categoría de accidente en autopista es:\n", u)

std = np.sqrt(mnb.var_)  # Desviación  para cada variable
print("La desviación estándar de las variables es:\n", std)

# Límites superior e inferior para cada variable en cada categoría
print("El límite superior de las variables es:\n", u + std)
print("El límite inferior de las variables es:\n", u - std)

La media de las variables por categoría de accidente en autopista es:
 [[3.88417722 7.45696203 0.86265823 4.35189873 0.66265823 3.21772152]
 [3.9261477  4.1497006  0.79840319 6.69461078 0.62075848 2.27744511]]
La desviación estándar de las variables es:
 [[1.95062542 3.52136997 0.34420782 2.3433645  0.47280262 1.87371962]
 [2.01603723 2.16198902 0.40119266 3.44163641 0.48519831 1.05765448]]
El límite superior de las variables es:
 [[ 5.83480263 10.978332    1.20686605  6.69526323  1.13546084  5.09144114]
 [ 5.94218494  6.31168962  1.19959585 10.13624719  1.10595679  3.33509959]]
El límite inferior de las variables es:
 [[1.9335518  3.93559205 0.51845041 2.00853424 0.18985561 1.3440019 ]
 [1.91011047 1.98771158 0.39721054 3.25297437 0.13556017 1.21979063]]


5. Evaluación del desempeño del modelo

In [ ]:
ydp=mnb.predict(X)  #Pronostico

from sklearn.metrics import confusion_matrix # se importa la función para calcular la matriz de confusión
cm=confusion_matrix(y,ydp) # Calcula la matriz de confusión
print (cm)

# Se extraen los valores de la matriz de confusión
VN = cm[0, 0]  # Verdaderos negativos
FP = cm[0, 1]  # Falsos positivos
FN = cm[1, 0]  # Falsos negativos
VP = cm[1, 1]  # Verdaderos positivos

Ex=(VP+VN)/(VP+VN+FP+FN) # Calcula y muestra la exactitud del modelo
print ("La exactitud del modelo es", Ex)

TE=(FP+FN)/(VP+VN+FP+FN) # Calcula y muestra el error del modelo
print ("El error del modelo es", TE)

Sen= VP/(VP+FN) # Calcula y muestra la sensibilidad del modelo
print ("La sensibilidad del modelo es", Sen)

ESP= VN/(VN+FP) # Calcula y muestra la especificidad del modelo
print ("La especificidad del modelo es", ESP)

pre= VP/(VP+FP) # Calcula y muestra la precisión del modelo
print ("La precisión del modelo es", pre)

pren= VN/(VN+FN) # Calcula y muestra la precisión negativa del modelo
print ("La precisión negativa del modelo es", pren)

[[1365  215]
 [ 224  277]]
La exactitud del modelo es 0.7890437289764536
El error del modelo es 0.21095627102354636
La sensibilidad del modelo es 0.5528942115768463
La especificidad del modelo es 0.8639240506329114
La precisión del modelo es 0.5630081300813008
La precisión negativa del modelo es 0.8590308370044053


**Análisis de resultados sobre el modelo:**
Según los resultados obtenidos por el modelo, se pudo verificar el rendimiento del modelo Naive Bayes utilizando el archivo de  "Car Crash ". Se observó que el modelo logró una precisión del 78.9%, lo que sugiere un buen desempeño general en la clasificación de accidentes automovilísticos tanto en autopistas como fuera de ellas.

Es especialmente notable el rendimiento del modelo en la predicción de accidentes fuera de autopistas, con una especificidad del 86.3%, lo que demuestra la alta capacidad para identificar correctamente estos casos. Además, la precisión para esta categoría alcanzó un 85.9%, lo que refuerza la fiabilidad del modelo en este caso

En general, los resultados muestran que el modelo tiene un buen desempeño al predecir accidentes fuera de las autopistas, aunque hay oportunidades para mejorar la detección de accidentes dentro de las autopistas.

El análisis del modelo Naive Bayes para predecir accidentes en autopistas en San Francisco indica que las variables predictoras empleadas no presentan diferencias significativas en sus valores promedio entre los accidentes ocurridos en autopistas y aquellos fuera de ellas, lo que sugiere que ninguna variable, por sí sola, actúa como un predictor fuerte. Sin embargo, es fundamental analizar la matriz de confusión y las métricas de evaluación (exactitud, precisión, sensibilidad y especificidad) para evaluar el rendimiento real del modelo.

**Respuestas preguntas**
1) ¿Cuál es el número de datos total de la base de datos?

In [ ]:
#Ya fue calculado en el punto 3 pero aqui lo pongo nuevamente
total_datos = len(X)
print(f"El número total de datos es: {total_datos}")

El número total de datos es: 2081


2) ¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?

In [ ]:
#También fue calculado en el punto 3 pero aqui lo pongo nuevamente
porcentajes_clase = ct / total_datos * 100
print("Porcentaje de datos en cada categoría:", porcentajes_clase)

Porcentaje de datos en cada categoría: [75.92503604 24.07496396]


3) Determinar los intervalos en los cual se encuentra la variación de la ViolCat para cada una de las categorías de clasificación

In [ ]:
std = np.sqrt(mnb.var_)  # Desviación  para cada variable- ViolCat es la N.2
print("La desviación estándar de las variables es:\n", std)

El intervalo de variación para el día de la semana es: (array([ 5.83480263, 10.978332  ,  1.20686605,  6.69526323,  1.13546084,
        5.09144114]), array([1.9335518 , 3.93559205, 0.51845041, 2.00853424, 0.18985561,
       1.3440019 ]))


4) Determinar el intervalo en los cuales se encuentra la variación de la Categoría de accidentes para cada una de las categorías de clasificación

In [ ]:
std = np.sqrt(mnb.var_)  # Desviación  para cada variable
print("La desviación estándar de las variables es:\n", std)

5) ¿Cuál es la exactitud alcanzada por el modelo?

In [ ]:
#Ya se encontró en el punto 5, pero aqui la dejo de nuevo
Ex=(VP+VN)/(VP+VN+FP+FN)
print ("La exactitud del modelo es", Ex)

6) ¿Cuál es la Especificidad alcanzada por el modelo?



In [ ]:
#También se encontró en el punto 5, pero aqui la dejo de nuevo
ESP= VN/(VN+FP)
print ("La especificidad del modelo es", ESP)

7) Indicar el intervalo en donde se ubica la cantidad de verdaderos negativo

In [ ]:
print(f"El valor de verdaderos negativos es: {VN}")

El valor de verdaderos negativos es: 1365


8) Indicar el intervalo en donde se ubica la cantidad de verdaderos positivos

In [ ]:
print(f"El valor de verdaderos positivos es: {VP}")

El valor de verdaderos positivos es: 277


9) ¿Cuál es la tasa de error alcanzada por el modelo?

In [ ]:
#También se encontró en el punto 5, pero aqui la dejo de nuevo
TE=(FP+FN)/(VP+VN+FP+FN)
print ("El error del modelo es", TE)

El error del modelo es 0.21095627102354636


10) ¿Cuál es la Sensibilidad alcanzada por el modelo?

In [ ]:
#También se encontró en el punto 5, pero aqui la dejo de nuevo
Sen= VP/(VP+FN)
print ("La sensibilidad del modelo es", Sen)

11) ¿Cuál es la Precisión Alcanzada por el Modelo?

In [ ]:
#También se encontró en el punto 5, pero aqui la dejo de nuevo
pre= VP/(VP+FP)
print ("La precisión del modelo es", pre)

12) ¿Indicar la categoría a la que pertenece una persona con las siguientes características [5,1,1,3,4,0]?

In [ ]:
#Ya se encontró en el punto 3 pero aqui lo dejo de nuevo
XS=[[5,1,1,3,4,0]]
XS= mnb.predict(XS)
print(XS)

[0]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(
